### DISCUSSION TITLE
2023-05-19 GPT Cheat Sheet (Dev Edition) - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#5577] [Chewing the Fat]
Hey team,

Let's talk about the [Dev Edition](https://www.ssw.com.au/rules/chatgpt-cheat-sheet/) cheat sheet.

**Comment by jernejk** [comment]
This is a great starting point for devs, I still use some of these, especially the refactoring and documentation ones. 😊

**Comment by jimmidier** [comment]
A tailored reference for devs, appreciate the work! 
I actually found it more helpful to use GPT when it comes to a very specific issue that involves complex contexts, in which case I need to clarify the tools I was using, the process I experienced to run the code, the system environment I was working on and so on. It's very similar to how a GitHub issue should be well described, and giving answers to such questions requires comprehensive knowledge that's nearly impossible to find on a single series of document, which is exactly when GPT comes in handy.

**Comment by tiagov8** [comment]
It's great content!

In terms of design, I think:
- The Tips section should have a different style from Prompt Templates.
- Not a big fan of the term "Prompt Templates" being repeated many times.
- Idon't see a good reason for having "examples" separated from "templates". I think templates are enough
- Less is more - Can we reduce the amount of text?

**Comment by matt-goldman** [comment]
Once upon a time, in the not-so-distant past of 2022, I discovered the greatest invention known to mankind, a product so innovative and revolutionary it put sliced bread to shame. This invention was none other than the legendary ChatGPT.

ChatGPT was a miracle of technology, an amalgamation of futuristic design and user-friendly functionality. Be it a morning routine or a midnight snack run, ChatGPT had me covered. So enamored was I with its efficiency and innovation, I quickly incorporated it into my everyday life. It became an essential aspect of my routine, as crucial as my morning cup of coffee.

However, as is often the case with fantastic discoveries, my well-kept secret soon found its way into the mainstream. It was like a hidden, underground band I had discovered becoming an overnight sensation. Suddenly, everyone was using ChatGPT, and they wouldn't stop talking about it. The world had caught the ChatGPT bug, and it was spreading faster than a high school rumor.

It was on every radio station, every news channel, and every social media platform. My friends would not stop raving about how it transformed their lives. "Have you heard of ChatGPT?" became a greeting. I couldn't turn a corner without seeing a ChatGPT CTF or overhearing a conversation about its latest feature. The world had transformed into an echo chamber of ChatGPT praise, and I was trapped inside it.

My once-beloved ChatGPT had transformed into a ubiquitous presence, and in an ironic twist of fate, I began to grow weary of it. The product I had once found refreshing and novel was now just another overhyped trend. It was like hearing my favorite song so many times that it lost its charm.

So one day, after hearing my barista suggest using ChatGPT to recommend my coffee order for the umpteenth time, I decided I had had enough. It was the final straw. ChatGPT, which had once held a special place in my daily routine, was ceremoniously retired. I declared myself a ChatGPT-free zone, much to the shock and horror of my ChatGPT-obsessed peers.

However, my life didn't stop. In fact, it kept going just as fine without ChatGPT. I found alternatives, returned to some good old-fashioned methods, and life went on. The best part? I didn't have to hear about ChatGPT from myself!

And so, the legend of ChatGPT lives on, a tale of love, obsession, and eventually, a break-up. It's the story of how something so loved can become so overwhelming, and how the phrase "too much of a good thing" is sometimes more real than we realize.

**Comment by ozdemir-mehmet** [comment]
Prompt: generate me a cheat sheet on using chatgpt as a software developer, give multiple language examples
Reponse: ... good not great.
Prompt: not for using ChatGPT directly but a cheatsheet for how to use ChatGPT to solve common software development problems
Response: Boom! Cheatsheat done.

**Comment by leehawkinsssw** [comment]
There's a lot of content in the dev cheat sheet, perhaps too much for one sheet.

It's a shame we don't have someone in SSW who's devoted their career to testing and could have provided better prompts in the Testing section of the sheet.

**Reply by bradystroud** [reply]
🤔

**Reply by 2fernandez** [reply]
**Re: too much content** - I could not have put this better myself. When you try to focus on everything, the main focus is lost.
**Testing profesional** - 🤣

**Comment by jakebayliss** [comment]
Agree with the above comments, I think it is a bit too busy and not sure about the dark mode for this specifically

**Comment by tkapa** [comment]
I liked the suggestion of making it do monotonous legwork for me.

**Comment by bahjatalaadelssw** [comment]
I think is an excellent tool to get quick responses to operation based question. The layout is clear and I doubt I'll have to reference the cheatsheet under duress, so I can take my time to skim through it. 😊

**Comment by vladislav-kir** [comment]
Chat GPT is very useful for documenting the code

**Comment by cvionnet** [comment]
It's a great addition to the general cheat sheet👍

**Comment by danielmackay** [comment]
Finally a proper cheat sheet targeting at developers!  I find this to be the most useful out of all the cheat sheets created recently.  It provides a wealth of examples that prompt (pun intended) usages of ChatGPT in different scenarios.  

Well done @bradystroud and team!

**Comment by duanxinhuan** [comment]
Great work!

**Comment by jeoffreyfischer** [comment]
The video was very useful for the coding part, however, the question regarding the c# documentation can be answered directly by google.

**Comment by calumjs** [comment]
I have seen a lot of GPT cheat sheets, and this is one of the better ones.
However, I think the concept of a cheat sheet is not that useful.
What is useful is understanding the principles, which you can do by playing around with it.

**Comment by KahaMason** [comment]
I think the cheatsheets work great hand-in-hand. The Dev cheatsheet is a great way for Devs up and running with  ChatGPT quickly.

**Comment by ncn-ssw** [comment]
The Dev Edition cheatsheet is good with a useful level of detail and a good layout for its level of detail.

**Comment by AshAnil** [comment]
The dev edition cheat sheet is really useful and saves a lot of time.

**Comment by wicksipedia** [comment]
This is nice and all, but given that I have access to CoPilot chat, I don’t see the need for it

1. Why look at a cheat sheet when GitHub Copilot can autocomplete my thoughts? 😏
2. Does a cheat sheet know my coding style like Copilot does? I think not! 🤨
3. I don't need a cheat sheet. I have an AI assistant that practically codes for me. 💁‍♂️
4. Copilot helps me write code faster than I can search a cheat sheet. ⚡
5. Who needs a cheat sheet when Copilot can suggest whole blocks of code? 😎
6. Cheat sheets can't help me explore new coding solutions. Copilot can! 🚀
I'd rather have an AI that learns from millions of repositories than a static cheat sheet. 🧠
7. Let's be real, using Copilot just makes me feel like I'm living in the future. 🛸
8. Copilot can look at my code and suggest your unit tests.

Please note that while these points are made in a lighthearted and humorous manner, they do reflect some of the actual capabilities and benefits of using GitHub Copilot. It's a powerful tool that can make coding more efficient and even help explore new solutions

**Comment by tomek-i** [comment]
🤖
**A**rtificial Intelligence, like ChatGPT, can greatly benefit software developers in their daily duties.
**B**y automating repetitive tasks, it saves them time and increases productivity.
**C**hatGPT can assist in generating code snippets, providing quick solutions to common problems.
**D**evelopers can leverage its vast knowledge base to learn new concepts and stay updated with the latest trends.
**E**fficiency is improved as ChatGPT helps in debugging and troubleshooting code errors.
**F**lexibility is enhanced as it can adapt to different programming languages and frameworks.
**G**enerating creative ideas and brainstorming becomes easier with ChatGPT's assistance.
**H**andling complex algorithms and data structures is simplified with its computational capabilities.
**I**nstant feedback and suggestions help developers refine their code and improve its quality.
**J**oining the developer community, ChatGPT fosters collaboration and knowledge sharing.
**K**nowledge acquisition is facilitated as developers can ask questions and get accurate answers.
**L**anguage and syntax nuances are clarified, preventing common mistakes and reducing frustration.
**M**ultitasking is made possible by having ChatGPT handle smaller tasks, freeing up developers' time.
**N**ovel solutions and alternative approaches can be explored with its diverse range of perspectives.
**O**ptimization is achieved by leveraging ChatGPT's ability to analyze code and identify bottlenecks.
**P**roductivity is boosted as developers can focus on higher-level tasks, leaving routine work to AI.
**Q**uality assurance is enhanced as ChatGPT assists in automated testing and bug detection.
**R**esource management is optimized through AI-driven suggestions for memory and performance.
**S**peedy prototyping is facilitated by quickly generating code templates and building blocks.
**T**ime-consuming research is minimized as ChatGPT can provide relevant documentation and references.
**U**ncovering hidden patterns and insights in large datasets becomes more efficient with AI.
**V**ersatility is expanded as ChatGPT can assist developers in various domains and industries.
**W**orkflow automation streamlines development processes, increasing overall efficiency.
**X**enial, ChatGPT promotes a positive and helpful environment for developers.
**Y**outhful energy exudes from developers empowered by ChatGPT's support and expertise.
**Z**ealous enthusiasm arises as developers realize the incredible potential of AI in their daily work.

**Comment by AlexRothwell** [comment]
The dev cheat sheet is so dense, it would likely benefit from being split into multiple cheat sheets based on the use case. These could then include examples of input and output so you could also search by the kind of output you're after

**Comment by william-liebenberg** [comment]
I think the cheat sheet could be a static site with light ☀️ and dark 🌙 mode toggle.

This way we can PR our own cheats to it :)

👍👍👍👍

**Comment by Marxoz** [comment]
A little bit much for one page. But as far as content goes, there are a few prompts that I did not think off when using ChatGPT. 

Maybe breaking it up into multiple cheat sheets will do the trick. Also having tech specific ones, like Azure cheat sheet, Bicep etc

Can we pop these into RuleGPT?

**Comment by MattParkerDev** [comment]
Great work on the dev cheat sheet, team! It's fantastic to see a comprehensive resource tailored specifically for developers. The cheat sheet covers a wide range of topics and provides valuable examples and prompts. I appreciate the effort put into the refactoring and documentation sections, as those are particularly useful for me.

I also wanted to mention how I've found GPT to be incredibly helpful in solving complex coding issues. Sometimes, a problem requires a deep understanding of the context, including the tools used, the code execution process, and the system environment. These are aspects that can be challenging to find in a single document, but that's where GPT comes in handy. It acts as a comprehensive knowledge base, allowing me to ask specific questions and receive accurate and relevant answers. It's like having a GitHub issue described in detail and being able to find the answers I need.

Overall, great content and a valuable resource for developers. Keep up the excellent work!

**Comment by Harry-Ross** [comment]
I think that the cheat sheet is a great way for developers to get up to speed with how they can use GPT to speed up development. The fact that it is a wide, general overview makes it really easy for any level of dev to easily see how ti can be used well. Really, just a great resource to have.

**Comment by SamBWagner** [comment]
I think the second one is more useful. The first was too general and didn't offer very much. 

I think we should get it made into a deskmat! it'd be a cool lil' piece of merch

**Comment by 2fernandez** [comment]
The Dev edition for Devs might be awesome for Devs.

But can someone please create a ChatGPT cheat sheet on:
"how to make my wife happy?" for the rest of us mortals?

![wife-romance-strangle](https://github.com/SSWConsulting/SSW.Rules.Content/assets/27670973/7c67cb41-b511-4909-a58c-5080ad4bd6a3)

**Reply by NadeekaJK** [reply]
Do you WANT to crash GPT with a fatal error?

**Comment by tombui99** [comment]
The dev cheat sheet is really useful and the code generation example from the video is definitely something I would use ChatGPT for in the future

**Comment by stephanfakossw** [comment]
I can see how the dev edition cheat sheet is a great tool for devs 😊📖

**Comment by AntPolkanov** [comment]
I like the dev cheat sheet more than the general one. Would love to see some of those features to be integrated/implemented in Rider.

**Comment by zacharykeeping** [comment]
Good video, Luke. The dev cheat sheet is good but I agree it seems a little busy and dense with information.

**Comment by amankumarrr** [comment]
Helpful tips, it saves a lot of time writing HTML code repeatedly or big JSON files.

**Comment by christoment** [comment]
I like the dev cheatsheet because it gives new ideas on how you can use ChatGPT 😄

**Comment by chrisschultzssw** [comment]
Looks good, I'll use some of these tips next time I need to write a script

**Comment by leahy268** [comment]
These are good starting points.  Very similar to what I already do but good as a quick check

**Comment by JackDevAU** [comment]
I agree with comments above, I think the Dev edition has a lot of valuable information, but I want to be able to copy and paste it. I think @william-liebenberg idea about creating a static website would be a lot better.

**Comment by JeanThirion** [comment]
🤖🤖🤖

Writing Code:

Instead of specifying the programming language (e.g., C#), use a more general approach that covers multiple languages (e.g., "Write a function to do {{ TASK }}").
Instead of specifying a specific framework or tool (e.g., React, Docker), consider using a more general approach to cover a wider range of technologies.
Debugging / Testing:

Consider providing more information about the error, such as the error message or stack trace, to help with troubleshooting.
Include information about the development environment (e.g., operating system, IDE) as it can be relevant for debugging.
Understanding Code:

Specify the context or purpose of the code to help the model provide a more accurate explanation (e.g., "What does this code do in the context of image processing?").
Instead of asking about the differences between two concepts, consider asking for an explanation or example of each concept separately.
Documenting Code:

Provide additional details or requirements for the documentation task (e.g., format, sections) to ensure more specific and relevant responses.
Consider including a specific focus for the documentation task, such as performance considerations or security guidelines.
Refactoring Code:

Provide more specific information about the desired outcome or goals of the refactoring task to receive more targeted suggestions.
Code Review:

Instead of a general request for code review, consider specifying the particular aspects of the code that require review (e.g., performance, maintainability).
By making these improvements, you can provide clearer and more focused prompts, which can help you receive more accurate and helpful responses from ChatGPT.

**Comment by PennyWalker** [comment]
I like the design of the first one, because it is clean and easy to read, but the dev cheat sheet has amazing content. It's too busy and needs to be broken up into smaller cheat sheets that are topic specific.

**Comment by MarcusIrmscherSSW** [comment]
Love the RulesBot, this is such a useful tool. I'm a fan of the normal Cheat sheet, the Dev version is a little too specialised for me.

**Comment by NadeekaJK** [comment]
I know the dev sheet has a lot of info and might seem busy.... but I like it when you have all the info in one place.

**Comment by bradystroud** [comment]
Dev cheat sheet looks like a real cheat sheet

**Comment by Ben0189** [comment]
The dev cheat sheet provide me new idea on how to using ChatGPT for coding

**Comment by Dhruv-0987** [comment]
Hey guys great cheat sheet, awesome content on there but one thing I would like is to specify the version of the library or the framework because often it gives code or method being used in the which are deprecated.

**Comment by CoffeeBugHub** [comment]
I find the Developer cheat sheet particularly valuable as it offers a wide range of prompt templates that are highly useful for day-to-day tasks in development work.

**Comment by UlyssesMaclaren** [comment]
These were really cool to read. So interesting to see how it can be used in a specifically developer setting. Made me want to see one for every job role!!! or every industry! Let's take over the world with ALL THE CHEAT SHEETS!

**Comment by KostaMadorsky** [comment]
Gonna try it out.

**Comment by Abf09** [comment]
Something that'll save me a lot of time and direct me in the right direction! Having it all together is very convenient.

**Comment by camillars** [comment]
I think both cheat sheets are practical, and everything will always have room for improvement :)

**Comment by drwharris** [comment]
Someone needs to pay my psych bills for all this GPT stuff!

**Comment by Hona** [comment]
while it might be good to show what you can ask chatgpt I think if you need to use it, you know the question you want to ask

**Comment by johnxu16** [comment]
It is insightful to ask questions with the cheat sheet 🤘

**Comment by sylhuang** [comment]
I felt that most of the questions are trivial and not worth time looking for a cheatsheet template. I would rather go ahead and ask the question and tailor the prompt as I go.

**Comment by JaneYangssw** [comment]
Great rule! I'll try to incorporate GPT more in my development to improve programming efficiency.

**Comment by Aibono1225** [comment]
Love the cheat sheet! It gave me some ideas I had never thought of before.

**Comment by jerryluossw** [comment]
Great rule, I have already used some of them.
in addition, for create C# or other language class from JSON object, there are a lot of online tools, so I'd like to use online tool directly, not use ChatGPT

**Comment by tino-liu** [comment]
Cheatsheet is definitely great to use. Most of the time I just end up asking ChatGPT the right questions that makes sense. And as long as you know what you're asking, I'm pretty sure ChatGPT will respond with something good. Personally, love using ChatGPT for code optimization.

**Comment by hveraus** [comment]
Great cheatsheet. I'd love to print it out. Can I have a light version? 👀

**Comment by KristenHu** [comment]
Such a detailed sheet, it looks great to me. I'm considering setting it as my desktop background.

**Comment by AnkyChe** [comment]
Greate rules and great cheatsheet. I love the detailed cheatsheet.

**Comment by zl24622** [comment]
The cheatsheet is helpful for remind me something I can do with GhatGPT.  Will have a try.
The RulesGPT is great.

**Comment by ozair-ashfaque** [comment]
Great video and effort. I would like to have advanced version of Dev Cheat Sheet

**Comment by BrookJeynes** [comment]
Dev cheatsheet looks good

**Comment by PothieuG** [comment]
Always good to have a great cheat sheet, I'll probably will use it often !
Thanks !

**Comment by TobyChurches** [comment]
Awesome Video @lukecookssw!! Cheatsheet looks great, Ill definitely be using these in my work and Uni 😁

**Comment by Berber-Mama** [comment]
Cheat Sheet is pretty handy! Love it!

**Comment by sethdaily** [comment]
I like the detail that we get from stuffing lots of info into it. Look forward to trying some of the ideas even though I'm not a dev.

**Comment by jackreimers** [comment]
It will be very handy to have all the prompts from the dev cheat sheet in once place, even though it is a little busy
